### PR TITLE
 nixos/pam: avoid extra lines in pam files 

### LINKED ISSUE #145286: Avoid lots of empty lines in /etc/pam.d/ files
### Describe the bug

Basically redo #18580 after it was reverted in #18710.

### Steps To Reproduce
1. `cat` any file in /etc/pam.d/

### Expected behavior

The non-empty lines should be separated by at most one empty line.

### Example

```console
$ cat /etc/pam.d/cups 
# Account management.
account required pam_unix.so






# Authentication management.











auth sufficient pam_unix.so   likeauth try_first_pass




auth required pam_deny.so

# Password management.
password sufficient pam_unix.so nullok sha512







# Session management.
session required pam_env.so conffile=/nix/store/0iliy98dpm3kcmmqxd7dzg0kmv7ksqyz-pam-environment readenv=0

session required pam_unix.so


















```

### Notify maintainers

@bjornfor Would you be interested?

### Metadata
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.76, NixOS, 21.05.4044.e74894146a4 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.05.4044.e74894146a4, home-manager-21.05"`
 - channels(victor): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by l0b0**
###### Motivation for this change

Closes https://github.com/NixOS/nixpkgs/issues/145286

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested, as applicable (see below)
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`) (N/A)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### Test procedure (on staging this time)

```console
$ cat > config.nix <<'EOF'
> {
>   config = {
>     krb5.enable = true;
>     security = {
>       apparmor.enable = true;
>       pam = {
>         enableEcryptfs = true;
>         enableSSHAgentAuth = true;
>         makeHomeDir.skelDirectory = true;
>       };
>     };
>     services.sssd.enable = true;
>     users.motd = "";
>     virtualisation.lxc.lxcfs.enable = true;
>   };
> }
> EOF
$ git checkout upstream/staging
[…]
$ git show --oneline
22a4fcfd4e8 (HEAD, upstream/staging) Merge pull request #145239 from Stunkymonkey/development-pname-version-1
$ cat "$(nix-build nixos --include nixos-config=config.nix --attr 'config.environment.etc."pam.d/xlock".source')" > ~/xlock.pam.orig
[…]
$ git checkout refactor/avoid-extra-empty-lines-in-etc-pamd-files
$ git rev-list --count HEAD ^upstream/staging
1
$ cat "$(nix-build nixos --include nixos-config=config.nix --attr 'config.environment.etc."pam.d/xlock".source')" > ~/xlock.pam.new
[…]
$ diff --ignore-blank-lines ~/xlock.pam.{orig,new} && echo 'No difference'
No difference
$ cat ~/xlock.pam.new
# Account management.
account required pam_unix.so
account sufficient /nix/store/xk2wjp29yhbwm3s2zs7bay518pmjz1pc-sssd-2.6.0/lib/security/pam_sss.so
account sufficient /nix/store/jpd67ssw170d4cw3gwg3crx4a50srvq5-pam-krb5-4.10/lib/security/pam_krb5.so

# Authentication management.
auth required pam_unix.so   likeauth
auth optional /nix/store/1gb81pi3dv937vqgd5vh2rppbi356cmi-ecryptfs-111/lib/security/pam_ecryptfs.so unwrap
auth sufficient pam_unix.so   likeauth try_first_pass
auth sufficient /nix/store/xk2wjp29yhbwm3s2zs7bay518pmjz1pc-sssd-2.6.0/lib/security/pam_sss.so use_first_pass
auth [default=ignore success=1 service_err=reset] /nix/store/jpd67ssw170d4cw3gwg3crx4a50srvq5-pam-krb5-4.10/lib/security/pam_krb5.so use_first_pass
auth [default=die success=done] /nix/store/gdj3886xf3wrmbpjq3xb0qa4nnvib7fp-pam_ccreds-10/lib/security/pam_ccreds.so action=validate use_first_pass
auth sufficient /nix/store/gdj3886xf3wrmbpjq3xb0qa4nnvib7fp-pam_ccreds-10/lib/security/pam_ccreds.so action=store use_first_pass
auth required pam_deny.so

# Password management.
password sufficient pam_unix.so nullok sha512
password optional /nix/store/1gb81pi3dv937vqgd5vh2rppbi356cmi-ecryptfs-111/lib/security/pam_ecryptfs.so
password sufficient /nix/store/xk2wjp29yhbwm3s2zs7bay518pmjz1pc-sssd-2.6.0/lib/security/pam_sss.so use_authtok
password sufficient /nix/store/jpd67ssw170d4cw3gwg3crx4a50srvq5-pam-krb5-4.10/lib/security/pam_krb5.so use_first_pass

# Session management.
session required pam_env.so conffile=/etc/pam/environment readenv=0
session required pam_unix.so
session optional /nix/store/1gb81pi3dv937vqgd5vh2rppbi356cmi-ecryptfs-111/lib/security/pam_ecryptfs.so
session optional /nix/store/xk2wjp29yhbwm3s2zs7bay518pmjz1pc-sssd-2.6.0/lib/security/pam_sss.so
session optional /nix/store/jpd67ssw170d4cw3gwg3crx4a50srvq5-pam-krb5-4.10/lib/security/pam_krb5.so
session optional /nix/store/v9cwqh3r09400pyilq722320rkx0w94x-lxc-4.0.11/lib/security/pam_cgfs.so -c all
```

Caveat: The tests above are not exhaustive - some options are mutually exclusive and some options should be passed to pam.nix rather than in config.nix. Hopefully the above will make it easy for anyone to test their favourite PAM section :)

In summary, this gets rid of a bunch of (all?) the unnecessary empty lines in /etc/pam.d files, while keeping the empty lines between sections of the file and seemingly keeping the structure of each of the entries.

@ttuegel, @bjornfor.

### CODE REVIEW

**Comment by ck3d** [comment]
I would argue that the Nix expression is read more often than the output of the expression. From my point of view, the old code look more appealing to me.

File: nixos/modules/security/pam.nix [hunk 1]
```diff
@@ -370,156 +370,262 @@ let
...
-              "account [default=bad success=ok user_unknown=ignore] ${pkgs.sssd}/lib/security/pam_sss.so"}
-          ${optionalString config.krb5.enable
-              "account sufficient ${pam_krb5}/lib/security/pam_krb5.so"}
-          ${optionalString cfg.googleOsLoginAccountVerification ''
+        (
+          ''
+            # Account management.
+            account required pam_unix.so
+          '' +
+          optionalString use_ldap
```
Comment (Artturin) [inline nixos/modules/security/pam.nix:378]: `'' + optionalString use_ldap ''` instead of on separate lines

**Comment by Artturin** [comment]
commit name should be
`nixos/pam: avoid extra lines in pam files`

File: nixos/modules/security/pam.nix (hunk 1 above)
Comment (Artturin) [reply nixos/modules/security/pam.nix:378]: nope doesn't look better :/
```
      text = mkDefault ( ''
        # Account management.
        account required pam_unix.so
        '' + optionalString use_ldap ''
          account sufficient ${pam_ldap}/lib/security/pam_ldap.so
        '' + optionalString (config.services.sssd.enable && cfg.sssdStrictAccess==false) ''
          account sufficient ${pkgs.sssd}/lib/security/pam_sss.so
        '' + optionalString (config.services.sssd.enable && cfg.sssdStrictAccess) ''
          account [default=bad success=ok user_unknown=ignore] ${pkgs.sssd}/lib/security/pam_sss.so
        '' + optionalString config.krb5.enable ''
          account sufficient ${pam_krb5}/lib/security/pam_krb5.so
        '' + optionalString cfg.googleOsLoginAccountVerification ''
          account [success=ok ignore=ignore default=die] ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so
          account [success=ok default=ignore] ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_admin.so
        '' + ''
```

image

i think it just isn't possible to do these well

Comment (Artturin) [reply nixos/modules/security/pam.nix:378]: yours is maybe a *little* bit more readable than the old one

Comment (thiagokokada) [reply nixos/modules/security/pam.nix:378]: Maybe using a list and `(concatStringsSep "\n (filter "" []))"` would make the code better?

Comment (thiagokokada) [reply nixos/modules/security/pam.nix:378]: Not tested:

```diff
diff --git a/nixos/modules/security/pam.nix b/nixos/modules/security/pam.nix
index c180af75bb1..1f02c9867fb 100644
--- a/nixos/modules/security/pam.nix
+++ b/nixos/modules/security/pam.nix
@@ -1015,78 +1015,76 @@ in
     security.apparmor.includes."abstractions/pam" = let
       isEnabled = test: fold or false (map test (attrValues config.security.pam.services));
       in
-      lib.concatMapStringsSep "\n"
+      (lib.concatMapStringsSep "\n"
         (name: "r ${config.environment.etc."pam.d/${name}".source},")
-        (attrNames config.security.pam.services) +
-      ''
-      mr ${getLib pkgs.pam}/lib/security/pam_filter/*,
-      mr ${getLib pkgs.pam}/lib/security/pam_*.so,
-      r ${getLib pkgs.pam}/lib/security/,
-      '' +
-      optionalString use_ldap ''
-         mr ${pam_ldap}/lib/security/pam_ldap.so,
-      '' +
-      optionalString config.services.sssd.enable ''
-        mr ${pkgs.sssd}/lib/security/pam_sss.so,
-      '' +
-      optionalString config.krb5.enable ''
-        mr ${pam_krb5}/lib/security/pam_krb5.so,
-        mr ${pam_ccreds}/lib/security/pam_ccreds.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.googleOsLoginAccountVerification)) ''
-        mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so,
-        mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_admin.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.googleOsLoginAuthentication)) ''
-        mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so,
-      '' +
-      optionalString (config.security.pam.enableSSHAgentAuth
-                     && isEnabled (cfg: cfg.sshAgentAuth)) ''
-        mr ${pkgs.pam_ssh_agent_auth}/libexec/pam_ssh_agent_auth.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.fprintAuth)) ''
-        mr ${pkgs.fprintd}/lib/security/pam_fprintd.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.u2fAuth)) ''
-        mr ${pkgs.pam_u2f}/lib/security/pam_u2f.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.usbAuth)) ''
-        mr ${pkgs.pam_usb}/lib/security/pam_usb.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.oathAuth)) ''
-        "mr ${pkgs.oathToolkit}/lib/security/pam_oath.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.yubicoAuth)) ''
-        mr ${pkgs.yubico-pam}/lib/security/pam_yubico.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.duoSecurity.enable)) ''
-        mr ${pkgs.duo-unix}/lib/security/pam_duo.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.otpwAuth)) ''
-        mr ${pkgs.otpw}/lib/security/pam_otpw.so,
-      '' +
-      optionalString config.security.pam.enableEcryptfs ''
-        mr ${pkgs.ecryptfs}/lib/security/pam_ecryptfs.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.pamMount)) ''
-        mr ${pkgs.pam_mount}/lib/security/pam_mount.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.enableGnomeKeyring)) ''
-        mr ${pkgs.gnome3.gnome-keyring}/lib/security/pam_gnome_keyring.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.startSession)) ''
-        mr ${pkgs.systemd}/lib/security/pam_systemd.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.enableAppArmor)
-                     && config.security.apparmor.enable) ''
-        mr ${pkgs.apparmor-pam}/lib/security/pam_apparmor.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.enableKwallet)) ''
-        mr ${pkgs.plasma5Packages.kwallet-pam}/lib/security/pam_kwallet5.so,
-      '' +
-      optionalString config.virtualisation.lxc.lxcfs.enable ''
-        mr ${pkgs.lxc}/lib/security/pam_cgfs.so
-      '';
+          (attrNames config.security.pam.services)) +
+      (lib.concatStringsSep "\n" (lib.remove "" [
+        ''
+          mr ${getLib pkgs.pam}/lib/security/pam_filter/*,
+          mr ${getLib pkgs.pam}/lib/security/pam_*.so,
+          r ${getLib pkgs.pam}/lib/security/,
+        ''
+        (optionalString use_ldap "mr ${pam_ldap}/lib/security/pam_ldap.so,")
+        (optionalString config.services.sssd.enable "mr ${pkgs.sssd}/lib/security/pam_sss.so,")
+        (optionalString config.krb5.enable ''
+          mr ${pam_krb5}/lib/security/pam_krb5.so,
+          mr ${pam_ccreds}/lib/security/pam_ccreds.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.googleOsLoginAccountVerification)) ''
+          mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so,
+          mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_admin.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.googleOsLoginAuthentication)) ''
+          mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so,
+        '')
+        (optionalString (config.security.pam.enableSSHAgentAuth
+                         && isEnabled (cfg: cfg.sshAgentAuth)) ''
+          mr ${pkgs.pam_ssh_agent_auth}/libexec/pam_ssh_agent_auth.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.fprintAuth)) ''
+          mr ${pkgs.fprintd}/lib/security/pam_fprintd.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.u2fAuth)) ''
+          mr ${pkgs.pam_u2f}/lib/security/pam_u2f.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.usbAuth)) ''
+          mr ${pkgs.pam_usb}/lib/security/pam_usb.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.oathAuth)) ''
+          "mr ${pkgs.oathToolkit}/lib/security/pam_oath.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.yubicoAuth)) ''
+          mr ${pkgs.yubico-pam}/lib/security/pam_yubico.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.duoSecurity.enable)) ''
+          mr ${pkgs.duo-unix}/lib/security/pam_duo.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.otpwAuth)) ''
+          mr ${pkgs.otpw}/lib/security/pam_otpw.so,
+        '')
+        (optionalString config.security.pam.enableEcryptfs ''
+          mr ${pkgs.ecryptfs}/lib/security/pam_ecryptfs.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.pamMount)) ''
+          mr ${pkgs.pam_mount}/lib/security/pam_mount.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.enableGnomeKeyring)) ''
+          mr ${pkgs.gnome3.gnome-keyring}/lib/security/pam_gnome_keyring.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.startSession)) ''
+          mr ${pkgs.systemd}/lib/security/pam_systemd.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.enableAppArmor)
+                         && config.security.apparmor.enable) ''
+          mr ${pkgs.apparmor-pam}/lib/security/pam_apparmor.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.enableKwallet)) ''
+          mr ${pkgs.plasma5Packages.kwallet-pam}/lib/security/pam_kwallet5.so,
+        '')
+        (optionalString config.virtualisation.lxc.lxcfs.enable ''
+          mr ${pkgs.lxc}/lib/security/pam_cgfs.so
+        '')
+      ]));
   };

 }
```

Comment (Artturin) [reply nixos/modules/security/pam.nix:378]: its easy to test these with a flake
```
{
  inputs = {
    #nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    nixpkgs.url = "/home/artturin/nixgits/my-nixpkgs";
  };

  outputs = inputs@{ self, nixpkgs }: {

    nixosConfigurations.vm = nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      specialArgs = { inherit inputs; };
      modules = [
        ({ pkgs, ... }: {
          documentation.enable = false;
          users.mutableUsers = false;
          users.users.root = {
            password = "root";
          };
          users.users.user = {
            password = "user";
            isNormalUser = true;
            extraGroups = [ "wheel" ];
          };
        })
      ];
    };
    # So that we can just run 'nix run' instead of
    # 'nix build ".#nixosConfigurations.vm.config.system.build.vm" && ./result/bin/run-nixos-vm'
    defaultPackage.x86_64-linux = self.nixosConfigurations.vm.config.system.build.vm;
    defaultApp.x86_64-linux = {
      type = "app";
      program = "${self.defaultPackage.x86_64-linux}/bin/run-nixos-vm";
    };
  };
}
```
just `nix build` and the files will be in `./result/system/etc/pam.d`

Comment (thiagokokada) [reply nixos/modules/security/pam.nix:378]: Well, I posted more for the idea than being the actual solution. I don't know what you think if this is more readable or not (I quite like it).

Comment (Artturin) [reply nixos/modules/security/pam.nix:378]: would be nice if optionalString didn't return a empty line but i dont think thats possible
```
  optionalString =
    # Condition
    cond:
    # String to return if condition is true
    string: if cond then string else "";
```

Comment (thiagokokada) [reply nixos/modules/security/pam.nix:378]: You can always create something like:

```nix
maybeString = cond: string: if cond then string else null;
```

And do:
```
(concatStringsSep "\n (remove null [ ... ]))"
```

Instead

**Comment by l0b0** [comment]
> I would argue that the Nix expression is read more often than the output of the expression. From my point of view, the old code look more appealing to me.

Since there are many more users than developers, I would expect the opposite. But we should definitely work on getting the most appealing code before merging.

That said, this is closer to the approach taken further down in the same file. Consistency would be nice too.

**Comment by Artturin** [comment]
this doesn't need to target staging as there are not many rebuilds

**Comment by l0b0** [comment]
> this doesn't need to target staging as there are not many rebuilds

Oh, that's how it works? Rebased again, and thanks for the tip.

File: nixos/modules/security/pam.nix [hunk 2]
```diff
@@ -370,156 +370,262 @@ let
...
+              account sufficient ${pam_ldap}/lib/security/pam_ldap.so
+            '' +
+          optionalString (config.services.sssd.enable && cfg.sssdStrictAccess==false)
+            ''
+              account sufficient ${pkgs.sssd}/lib/security/pam_sss.so
+            '' +
+          optionalString (config.services.sssd.enable && cfg.sssdStrictAccess)
+            ''
+              account [default=bad success=ok user_unknown=ignore] ${pkgs.sssd}/lib/security/pam_sss.so
+            '' +
```
Comment (Artturin) [inline nixos/modules/security/pam.nix:389]:
```suggestion
          optionalString (config.services.sssd.enable && cfg.sssdStrictAccess==false) ''
              account sufficient ${pkgs.sssd}/lib/security/pam_sss.so
          '' +
          optionalString (config.services.sssd.enable && cfg.sssdStrictAccess) ''
              account [default=bad success=ok user_unknown=ignore] ${pkgs.sssd}/lib/security/pam_sss.so
          '' +
```

like this it matches whats down below

**Comment by Artturin** [comment]
are you familiar with pam?

**Comment by l0b0** [comment]
> are you familiar with pam? / do you have experience with it

Only a couple basic changes to /etc/pam.d files. I tried to make properly sure only empty lines were removed though, and my (limited) tests bear that out. I don't have flake support enabled, but if you have time to check using [your configuration](https://github.com/NixOS/nixpkgs/pull/145574#discussion_r748642001) that would be useful to make sure we don't deviate semantically.

File: nixos/modules/security/pam.nix (hunk 2 above)
Comment (l0b0) [reply nixos/modules/security/pam.nix]: Thanks! I've applied this suggestion to the whole thing.

**Comment by ck3d** [comment]
Your current version looks better than the original one, great!

**Comment by Artturin** [comment]
making a diff script..

**Comment by Artturin** [comment]
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/master";
  };

  outputs = inputs@{ self, nixpkgs }: {

    nixosConfigurations.vm = nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      specialArgs = { inherit inputs; };
      modules = [
        ({ pkgs, lib, modulesPath, ... }: {
          documentation.enable = false;
          users.mutableUsers = false;
          users.users.root = {
            password = "root";
          };
          users.users.user = {
            password = "user";
            isNormalUser = true;
            extraGroups = [ "wheel" ];
          };

          imports = [
            (modulesPath + "/profiles/minimal.nix")
          ];
        })
      ];
    };
    # So that we can just run 'nix run' instead of
    # 'nix build ".#nixosConfigurations.vm.config.system.build.vm" && ./result/bin/run-nixos-vm'
    defaultPackage.x86_64-linux = self.nixosConfigurations.vm.config.system.build.vm;
    defaultApp.x86_64-linux = {
      type = "app";
      program = "${self.defaultPackage.x86_64-linux}/bin/run-nixos-vm";
    };
  };
}
```

```
#!/usr/bin/env bash

nix build -o resultNoFix
nix build --override-input nixpkgs "github:l0b0/nixpkgs/refactor/avoid-extra-empty-lines-in-etc-pamd-files" -o resultFix

files=$(ls resultNoFix/system/etc/pam.d/)

for f in $files; do 
    diff --ignore-blank-lines $(chase resultNoFix/system/etc/pam.d/$f) $(chase resultFix/system/etc/pam.d/$f)
done
```
no output so there are no differences other than blank line changes

**Review by Artturin — APPROVED** [review]

**Comment by l0b0** [comment]
> no output so there are no differences other than blank line changes

Excellent! I don't really want to delay this any longer, but is there some way this could be made into a test? PAM is super important to get right, so it would be great to have some actual proof that it generates what we expect.

**Comment by Artturin** [comment]
> > no output so there are no differences other than blank line changes
> 
> Excellent! I don't really want to delay this any longer, but is there some way this could be made into a test? PAM is super important to get right, so it would be great to have some actual proof that it generates what we expect.

maybe creating a vm test which runs pamtester commands http://pamtester.sourceforge.net
and various other command

**Comment by Artturin** [comment]
rebuilt my system https://nixos.wiki/wiki/Nixpkgs/Reviewing_changes#Modules

```
  disabledModules = [ "security/pam.nix" ];
  
  imports = let
    pkgsReview = builtins.fetchTarball {
      url = "https://github.com/l0b0/nixpkgs/archive/refactor/avoid-extra-empty-lines-in-etc-pamd-files.tar.gz";
    };
  in [
    (import "${pkgsReview}/nixos/modules/security/pam.nix")
  ];
```

and everything seems ok

**Comment by l0b0** [comment]
> maybe creating a vm test which runs pamtester commands http://pamtester.sourceforge.net and various other command

I like the overall idea, I just wouldn't want to use a 15 year old tool hosted on an HTTP web site with a provider known for injecting stuff into the packages they host.
